File: php/src/Render.php
<?php

declare(strict_types=1);

namespace LiteDbModel\Runtime;

use LiteDbModel\Runtime\BehaviorContracts\ExprEval;

/**
 * litedbmodel v2 SCP — fragment-tree render + param assembly (PHP port of src/scp/render.ts, WS7d #33).
 *
 * The NORMATIVE render reference ported BYTE-FOR-BYTE from the TS `renderOperation`
 * (docs/proposal/sql-dynamic-expansion-spec.md). Given a §8 `CompiledOperation` (as a stdClass
 * decoded from the bundle JSON via json_decode(.., false)) and a bound input scope (assoc array),
 * it deterministically produces the final SQL text + flat params, semantics-identical to TS:
 *
 *   §2 SKIP → fragment existence (`when` truthy/present; absent = no SQL, no params)
 *   §3 empty-WHERE degeneration (no present fragment ⇒ ` WHERE ` keyword collapses)
 *   §4 AND/OR structure + parenthesization
 *   §5 IN-list array expansion (`(?)` → `(?, ?, …)`; empty ⇒ `1 = 0`)
 *   §6 param order = SQL text order (pre-`{where}` statics, fragment params, post statics)
 *   §8 dialect final placeholder pass (`?`→`$N` for Postgres, once over the assembled text)
 *
 * All Expression-IR value evaluation (param slots + SKIP `when` guards) is delegated to the
 * VENDORED behavior-contracts `ExprEval::evaluate` — the runtime re-implements NO expr eval.
 */
final class Render
{
    /** The literal `{where}` splice marker inside `CompiledOperation.sql` (ir.ts WHERE_SLOT). */
    public const WHERE_SLOT = '{where}';

    /**
     * Render a §8 compiled operation (stdClass) against a bound scope for a dialect.
     *
     * @param \stdClass $op    the §8 CompiledOperation (json_decode(.., false) shape).
     * @param array<string,mixed> $scope the bound input scope (bc flat scope).
     * @return array{sql:string, params:list<mixed>} final SQL + flat params (1:1 with `?`/`$N`).
     */
    public static function renderOperation(\stdClass $op, array $scope, Dialect $dialect): array
    {
        $params = [];
        $sqlText = (string) ($op->sql ?? '');
        $markerIdx = strpos($sqlText, self::WHERE_SLOT);

        if ($markerIdx === false) {
            // No dynamic WHERE: all params are static, in position order.
            foreach (self::opParams($op) as $slot) {
                $params[] = ExprEval::evaluate($slot, $scope);
            }
            return ['sql' => $dialect->finalizePlaceholders($sqlText), 'params' => $params];
        }

        $before = substr($sqlText, 0, $markerIdx);
        $after = substr($sqlText, $markerIdx + strlen(self::WHERE_SLOT));

        // Static params are partitioned by whether their `?` sits before or after the marker.
        $beforeQ = self::countPlaceholders($before);
        $allStatic = self::opParams($op);
        $preStatics = array_slice($allStatic, 0, $beforeQ);
        $postStatics = array_slice($allStatic, $beforeQ);

        foreach ($preStatics as $slot) {
            $params[] = ExprEval::evaluate($slot, $scope);
        }

        $whereSql = '';
        $where = $op->where ?? null;
        if ($where instanceof \stdClass) {
            $body = self::renderTree($where, $scope, $params);
            if ($body !== '') {
                $whereSql = ' WHERE ' . $body; // degeneration §3: drop keyword when empty
            }
        }

        foreach ($postStatics as $slot) {
            $params[] = ExprEval::evaluate($slot, $scope);
        }

        return [
            'sql' => $dialect->finalizePlaceholders($before . $whereSql . $after),
            'params' => $params,
        ];
    }

    /** The op's static param slots as a plain list (json array from the bundle). @return list<mixed> */
    private static function opParams(\stdClass $op): array
    {
        $p = $op->params ?? [];
        return is_array($p) ? array_values($p) : [];
    }

    /**
     * A fragment tree carries a `connector`; a fragment carries `sql` (render.ts `isTree`).
     */
    private static function isTree(\stdClass $node): bool
    {
        return property_exists($node, 'connector');
    }

    /**
     * SKIP existence rule (spec §2): present iff `always === true`, or `when` evaluates to a
     * PRESENT binding (`null`/`false` absent; everything else present). Fail-closed: neither
     * `always` nor `when` ⇒ absent. Mirrors render.ts `fragmentPresent`.
     *
     * @param array<string,mixed> $scope
     */
    private static function fragmentPresent(\stdClass $f, array $scope): bool
    {
        if (($f->always ?? null) === true) {
            return true;
        }
        if (!property_exists($f, 'when')) {
            return false; // fail-closed: neither always nor when
        }
        $v = ExprEval::evaluate($f->when, $scope);
        return $v !== null && $v !== false;
    }

    /**
     * Render one leaf fragment's SQL + params (render.ts `renderFragment`). Handles IN-list
     * expansion (spec §5): the `expand` slot is an array; its `(?)` becomes `(?, ?, …)` and each
     * element is pushed as its own param; empty array degenerates the whole fragment to `1 = 0`.
     *
     * @param array<string,mixed> $scope
     * @param list<mixed> $params accumulator (by ref)
     */
    private static function renderFragment(\stdClass $f, array $scope, array &$params): string
    {
        $slots = is_array($f->params ?? null) ? array_values($f->params) : [];
        $expand = property_exists($f, 'expand') ? $f->expand : null;
        $fSql = (string) ($f->sql ?? '');

        if ($expand === null) {
            foreach ($slots as $slot) {
                $params[] = ExprEval::evaluate($slot, $scope);
            }
            return $fSql;
        }

        // IN-list expansion. Evaluate all slots; the `expand` slot must be an array.
        $sql = $fSql;
        $n = count($slots);
        for ($i = 0; $i < $n; $i++) {
            $v = ExprEval::evaluate($slots[$i], $scope);
            if ($i === $expand) {
                if (!is_array($v)) {
                    $got = $v === null ? 'null' : gettype($v);
                    throw new \RuntimeException(
                        "IN-list expansion slot {$i} did not bind to an array (got {$got})"
                    );
                }
                if (count($v) === 0) {
                    // Empty-array degeneration (spec §5): `col IN (?)` collapses to `1 = 0`;
                    // NO params pushed for this slot (byte-identical to v1's empty-IN handling).
                    $sql = '1 = 0';
                } else {
                    $placeholders = implode(', ', array_fill(0, count($v), '?'));
                    $sql = self::replaceFirst($sql, '(?)', '(' . $placeholders . ')');
                    foreach ($v as $el) {
                        $params[] = $el;
                    }
                }
            } else {
                $params[] = $v;
            }
        }
        return $sql;
    }

    /**
     * Render a fragment tree into a WHERE clause body (no leading ` WHERE `). Present fragments
     * are joined by ` <connector> `; a nested tree is parenthesized (spec §4). Empty string when
     * NO fragment is present (§3). Mirrors render.ts `renderTree`.
     *
     * @param array<string,mixed> $scope
     * @param list<mixed> $params accumulator (by ref)
     */
    private static function renderTree(\stdClass $tree, array $scope, array &$params): string
    {
        $parts = [];
        $fragments = is_array($tree->fragments ?? null) ? $tree->fragments : [];
        foreach ($fragments as $node) {
            if (!($node instanceof \stdClass)) {
                continue;
            }
            if (self::isTree($node)) {
                $inner = self::renderTree($node, $scope, $params);
                if ($inner !== '') {
                    $parts[] = '(' . $inner . ')';
                }
            } elseif (self::fragmentPresent($node, $scope)) {
                $parts[] = self::renderFragment($node, $scope, $params);
            }
        }
        if (count($parts) === 0) {
            return '';
        }
        $connector = (string) ($tree->connector ?? 'AND');
        return implode(' ' . $connector . ' ', $parts);
    }

    /** Count `?` placeholders in a static SQL segment (render.ts `countPlaceholders`). */
    private static function countPlaceholders(string $sql): int
    {
        return substr_count($sql, '?');
    }

    /** Replace only the FIRST occurrence of `$needle` in `$haystack` (TS String.replace semantics). */
    private static function replaceFirst(string $haystack, string $needle, string $replace): string
    {
        $pos = strpos($haystack, $needle);
        if ($pos === false) {
            return $haystack;
        }
        return substr($haystack, 0, $pos) . $replace . substr($haystack, $pos + strlen($needle));
    }
}

File: go/litedbmodel_runtime/render.go
// litedbmodel v2 SCP — fragment-tree render + param assembly (Go port of src/scp/render.ts).
//
// The NORMATIVE render reproduced EXACTLY from the TS golden reference: given a §8
// CompiledOperation (parsed as bc JNode) and a bound scope, it deterministically produces the
// final SQL text (`?` placeholders, then the dialect's final placeholder pass) and the flat
// params slice. Every rule is byte-true to render.ts:
//
//	§2 SKIP → fragment existence (present iff `always` OR `when` evaluates to a present binding —
//	    null/false are absent).
//	§3 empty-WHERE degeneration (no present fragment ⇒ the whole `{where}` splice, WHERE keyword
//	    included, collapses to "").
//	§4 AND/OR structure + parenthesization for a nested tree.
//	§5 IN-list array expansion (`(?)` → `(?, ?, …)`; empty array → the `1 = 0` always-false
//	    degeneration, byte-identical to v1's empty-IN).
//	§6 param order = SQL text order (pre-WHERE statics, then fragment params, then post-WHERE).
//	§8 the `?`→`$N` conversion runs ONCE at the end (PG only), via the Dialect.
//
// The generic Expression-IR param/guard evaluation is delegated to the bc common core
// (EvaluateExpression) — this file re-implements NO evaluator, only the SQL-structural assembly.

package litedbmodel_runtime

import (
	"fmt"
	"strings"

	bc "github.com/foo-ogawa/behavior-contracts/go"
)

// whereSlot is the literal `{where}` splice marker inside CompiledOperation.sql (ir.ts WHERE_SLOT).
const whereSlot = "{where}"

// Rendered is the output of rendering one §8 CompiledOperation: final SQL text + flat params
// (1:1 with `?`/`$N`, each a bc runtime Value).
type Rendered struct {
	SQL    string
	Params []bc.Value
}

// asJObj narrows a parsed JNode to a *bc.JObj (nil if not an object).
func asJObj(n bc.JNode) (*bc.JObj, bool) {
	o, ok := n.(*bc.JObj)
	return o, ok
}

// isTree reports whether a fragment node is a FragmentTree (carries a `connector`).
func isTree(node *bc.JObj) bool {
	_, ok := node.Get("connector")
	return ok
}

// fragmentPresent applies the SKIP existence rule (render.ts fragmentPresent): present iff
// `always === true`, or `when` evaluates to a present binding (not null, not false). Fail-closed
// when neither `always` nor `when` is set.
func fragmentPresent(f *bc.JObj, scope *bc.Obj) (bool, error) {
	if a, ok := f.Get("always"); ok {
		if b, isBool := a.(bool); isBool && b {
			return true, nil
		}
	}
	whenN, hasWhen := f.Get("when")
	if !hasWhen {
		return false, nil // fail-closed: neither always nor when
	}
	v, err := bc.EvaluateExpression(whenN, scope)
	if err != nil {
		return false, err
	}
	if v == nil {
		return false, nil
	}
	if b, ok := v.(bool); ok && !b {
		return false, nil
	}
	return true, nil
}

// fragmentSQL returns the fragment's literal `sql` text.
func fragmentSQL(f *bc.JObj) (string, error) {
	s, ok := f.Get("sql")
	if !ok {
		return "", fmt.Errorf("scp render: fragment missing 'sql'")
	}
	str, ok := s.(string)
	if !ok {
		return "", fmt.Errorf("scp render: fragment 'sql' is not a string")
	}
	return str, nil
}

// fragmentParams returns the fragment's param-slot JNodes (1:1 with its `?`).
func fragmentParams(f *bc.JObj) []bc.JNode {
	p, ok := f.Get("params")
	if !ok {
		return nil
	}
	arr, ok := p.([]bc.JNode)
	if !ok {
		return nil
	}
	return arr
}

// fragmentExpand returns the IN-list expansion slot index (and whether it is set) (ir.ts `expand`).
func fragmentExpand(f *bc.JObj) (int, bool) {
	e, ok := f.Get("expand")
	if !ok {
		return 0, false
	}
	// The bundle encodes numbers as json.Number (bc ParseJSONOrdered UseNumber).
	switch n := e.(type) {
	case int64:
		return int(n), true
	default:
		// json.Number path: coerce via bc's own decode.
		if v, err := bc.DecodeValue(e); err == nil {
			if i, ok := v.(int64); ok {
				return int(i), true
			}
		}
	}
	return 0, false
}

// renderFragment renders one leaf fragment's SQL + params into the accumulator, handling IN-list
// expansion (render.ts renderFragment, byte-true incl. the `1 = 0` empty-array degeneration).
func renderFragment(f *bc.JObj, scope *bc.Obj, params *[]bc.Value) (string, error) {
	sql, err := fragmentSQL(f)
	if err != nil {
		return "", err
	}
	slots := fragmentParams(f)
	expandIdx, hasExpand := fragmentExpand(f)
	if !hasExpand {
		for _, slot := range slots {
			v, err := bc.EvaluateExpression(slot, scope)
			if err != nil {
				return "", err
			}
			*params = append(*params, v)
		}
		return sql, nil
	}
	// IN-list expansion. Evaluate all slots; the `expand` slot must be an array.
	for i, slot := range slots {
		v, err := bc.EvaluateExpression(slot, scope)
		if err != nil {
			return "", err
		}
		if i == expandIdx {
			arr, isArr := v.([]bc.Value)
			if !isArr {
				return "", fmt.Errorf("IN-list expansion slot %d did not bind to an array (got %s)", i, typeLabel(v))
			}
			if len(arr) == 0 {
				// Empty-array degeneration (spec §5): `col IN (?)` collapses to `1 = 0`; no params.
				sql = "1 = 0"
			} else {
				qs := make([]string, len(arr))
				for j := range arr {
					qs[j] = "?"
				}
				sql = strings.Replace(sql, "(?)", "("+strings.Join(qs, ", ")+")", 1)
				*params = append(*params, arr...)
			}
		} else {
			*params = append(*params, v)
		}
	}
	return sql, nil
}

// typeLabel mirrors the TS `null`/typeof message for the IN-list error text (best-effort parity).
func typeLabel(v bc.Value) string {
	if v == nil {
		return "null"
	}
	switch v.(type) {
	case bool:
		return "boolean"
	case int64, float64:
		return "number"
	case string:
		return "string"
	case *bc.Obj:
		return "object"
	default:
		return "object"
	}
}

// renderTree renders a fragment tree into a WHERE clause body (no leading ` WHERE `). Present
// fragments join by ` <connector> `; a nested tree is parenthesized. Empty when no fragment is
// present (render.ts renderTree).
func renderTree(tree *bc.JObj, scope *bc.Obj, params *[]bc.Value) (string, error) {
	connector, _ := tree.Get("connector")
	conn, _ := connector.(string)
	fragsN, _ := tree.Get("fragments")
	frags, _ := fragsN.([]bc.JNode)

	var parts []string
	for _, node := range frags {
		fo, ok := asJObj(node)
		if !ok {
			return "", fmt.Errorf("scp render: fragment node is not an object")
		}
		if isTree(fo) {
			inner, err := renderTree(fo, scope, params)
			if err != nil {
				return "", err
			}
			if inner != "" {
				parts = append(parts, "("+inner+")")
			}
			continue
		}
		present, err := fragmentPresent(fo, scope)
		if err != nil {
			return "", err
		}
		if present {
			frag, err := renderFragment(fo, scope, params)
			if err != nil {
				return "", err
			}
			parts = append(parts, frag)
		}
	}
	if len(parts) == 0 {
		return "", nil
	}
	return strings.Join(parts, " "+conn+" "), nil
}

// countPlaceholders counts `?` in a static SQL segment (render.ts countPlaceholders).
func countPlaceholders(sql string) int {
	return strings.Count(sql, "?")
}

// operationSQL / operationParams / operationWhere / operationComponent read the compiled-op fields.
func operationSQL(op *bc.JObj) (string, error) {
	s, ok := op.Get("sql")
	if !ok {
		return "", fmt.Errorf("scp render: operation missing 'sql'")
	}
	str, ok := s.(string)
	if !ok {
		return "", fmt.Errorf("scp render: operation 'sql' is not a string")
	}
	return str, nil
}

func operationParams(op *bc.JObj) []bc.JNode {
	p, ok := op.Get("params")
	if !ok {
		return nil
	}
	arr, _ := p.([]bc.JNode)
	return arr
}

func operationWhere(op *bc.JObj) *bc.JObj {
	w, ok := op.Get("where")
	if !ok {
		return nil
	}
	o, _ := w.(*bc.JObj)
	return o // nil when where is JSON null
}

func operationComponent(op *bc.JObj) string {
	c, ok := op.Get("component")
	if !ok {
		return ""
	}
	s, _ := c.(string)
	return s
}

// RenderOperation renders a §8 CompiledOperation (parsed JNode) to final SQL + params for a bound
// scope and dialect. Ported EXACTLY from render.ts renderOperation (pre/post-WHERE static
// partition, fragment splice, then the dialect's finalize pass).
func RenderOperation(op *bc.JObj, scope *bc.Obj, dialect Dialect) (Rendered, error) {
	var params []bc.Value
	sql, err := operationSQL(op)
	if err != nil {
		return Rendered{}, err
	}
	staticSlots := operationParams(op)

	markerIdx := strings.Index(sql, whereSlot)
	if markerIdx == -1 {
		// No dynamic WHERE: all params are static, in position order.
		for _, slot := range staticSlots {
			v, err := bc.EvaluateExpression(slot, scope)
			if err != nil {
				return Rendered{}, err
			}
			params = append(params, v)
		}
		return Rendered{SQL: dialect.FinalizePlaceholders(sql), Params: params}, nil
	}

	before := sql[:markerIdx]
	after := sql[markerIdx+len(whereSlot):]

	// Static params partitioned by whether their `?` sits before or after the marker.
	beforeQ := countPlaceholders(before)
	var preStatics, postStatics []bc.JNode
	if beforeQ <= len(staticSlots) {
		preStatics = staticSlots[:beforeQ]
		postStatics = staticSlots[beforeQ:]
	} else {
		preStatics = staticSlots
	}

	for _, slot := range preStatics {
		v, err := bc.EvaluateExpression(slot, scope)
		if err != nil {
			return Rendered{}, err
		}
		params = append(params, v)
	}

	whereSQL := ""
	if where := operationWhere(op); where != nil {
		body, err := renderTree(where, scope, &params)
		if err != nil {
			return Rendered{}, err
		}
		if body != "" {
			whereSQL = " WHERE " + body // degeneration §3: drop the keyword when empty
		}
	}

	for _, slot := range postStatics {
		v, err := bc.EvaluateExpression(slot, scope)
		if err != nil {
			return Rendered{}, err
		}
		params = append(params, v)
	}

	return Rendered{SQL: dialect.FinalizePlaceholders(before + whereSQL + after), Params: params}, nil
}
